src/mlpack/methods/ann/layer/group_norm_impl.hpp

```diff
@@ -0,0 +1,196 @@
+/**
+ * @file methods/ann/layer/group_norm_impl.hpp
+ * @author Abhinav Anand
+ *
+ * Implementation of the Group Normalization class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_LAYER_GROUPNORM_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_GROUPNORM_IMPL_HPP
+
+// In case it is not included.
+#include "group_norm.hpp"
+
+namespace mlpack {
+namespace ann { /** Artificial Neural Network. */
+
+
+template<typename InputDataType, typename OutputDataType>
+GroupNorm<InputDataType, OutputDataType>::GroupNorm() :
+    channelSize(0),
+    size(0),
+    groupCount(1),
+    eps(1e-8),
+    loading(false)
+{
+  // Nothing to do here.
+}
+
+template <typename InputDataType, typename OutputDataType>
+GroupNorm<InputDataType, OutputDataType>::GroupNorm(
+    const size_t channelSize, const size_t size, const size_t groupCount, const double eps) :
+    channelSize(channelSize),
+    size(size),
+    groupCount(groupCount),
+    eps(eps),
+    loading(false)
+{
+  if (size % groupCount != 0)
+  {
+    Log::Fatal << "Total input units must be divisible by groupCount!" << std::endl;
+  }
+
+  weights.set_size(size * 2 + 1, 1);
+}
+
+template<typename InputDataType, typename OutputDataType>
+void GroupNorm<InputDataType, OutputDataType>::Reset()
+{
+  gamma = arma::mat(weights.memptr(), size, 1, false, false);
+  beta = arma::mat(weights.memptr() + gamma.n_elem, size, 1, false, false);
+
+  if (!loading)
+  {
+    gamma.fill(1.0);
+    beta.fill(0.0);
+  }
+
+  loading = false;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void GroupNorm<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>& input, arma::Mat<eT>& output)
+{
+  if (output.is_empty())
+    output.set_size(input.n_rows / groupCount, input.n_cols * groupCount);
+
+  assert(input.n_rows % channelSize == 0);
+  assert(input.n_rows % size == 0);
+
+  arma::mat reshapedInput(const_cast<arma::Mat<eT>&>(input).memptr(),
+      input.n_rows / groupCount, input.n_cols * groupCount, false, false);
+
+  mean = arma::mean(reshapedInput, 0);
+  variance = arma::var(reshapedInput, 1, 0);
+
+  // Normalize the input.
+  output = reshapedInput.each_row() - mean;
+  inputMean = output;
+  output.each_row() /= arma::sqrt(variance + eps);
+
+  output.reshape(input.n_rows, input.n_cols);
+  // Reused in the backward and gradient step.
+  normalized = output;
+
+  arma::mat expandedGamma, expandedBeta;
+  expandedGamma.set_size(input.n_rows, 1);
```

Comment:
Nice catch with the rows.

Reply:
@zoq    The test case  is still failing for some reason. But in my local build its passing.